### PR TITLE
MAINT: stats: batch permutation t-test

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Supersedes gh-14206
gh-4824
gh-13371
gh-14126

#### What does this implement/fix?
To reduce memory consumption, this processes the permutation t-test in batches of size 50 rather than all permutations at once. This was considered in gh-4824, but not implemented initially for simplicity. gh-13371 highlighted this feature as important for the bootstrap, so we pushed to address this in gh-14126 before 1.7 branched. This PR didn't make it in time for branching 1.7 or 1.8, but I'll let others decide whether it is important enough to backport, as excessive memory consumption could be considered to be a bug.

#### Additional information
In the past, a batch size of 50-100 has been the sweet spot on my machine. I chose the lower end here to reduce memory consumption, and I hard-coded the value to avoid complicating the t-test interface further. If the user wants a fully customizable permutation test, there the more general `permutation_test` function is already available (gh-13899).

### CODE REVIEW

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6287]:
```suggestion
    """Generation permutation distribution of t statistic"""
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6071]: I thought I should add this note based on what I learned finishing up gh-13899 (see [here](https://github.com/scipy/scipy/pull/13899#issuecomment-986104041)). You can check the accuracy of this note against the code by comparing the Lines 6393-6399 against the note linked above and also the Efron's definition of the two-sided ASL: 
 
image

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:37]:
```suggestion
from scipy.stats._stats_py import (_permutation_distribution_t,
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6069]:
```suggestion
    (ASL) as defined in Chapter 15 of [3]_ but with a non-strict inequality.
    Note that there are other ways of
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6071]:
```suggestion
    (ASL) as defined in Chapter 15 of [3]_ but with a non-strict inequality.
    Note that there are other ways of estimating p-values using 
    randomized permutation tests; for other options,
    see the more general `permutation_test`.
```

Comment (steppi) [reply scipy/stats/_stats_py.py:6071]: For what it’s worth, in Efron’s more recent book, [Computer Age Statistical Inference](https://hastie.su.domains/CASI_files/PDF/casi.pdf), this test is defined with `>=`. 
9FAC3ACE-2C5B-4513-9540-C2CC27D17BA2

Comment (mdhaber) [reply scipy/stats/_stats_py.py:6071]: Much better! I'll list that as the reference.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6094]:
```suggestion
    .. [3] B. Efron and T. Hastie. Computer Age Statistical Inference. (2016).
```

File: scipy/stats/_stats_py.py
Comment (rlucas7) [reply scipy/stats/_stats_py.py]: FWIW in large samples the distinction with the equality will usually be irrelevant-assuming I'm understanding the context correctly. The difference would matter in small sample sizes and you might not be able to achieve the exact coverage probabilities without an additional layer of randomization. When you add the additional layer of randomization these are known as randomized hypothesis tests-you flip a coin w/probability heads =p where p is set to achieve the type I error probability you want or specified. I don't think it makes sense in this PR to include randomization though documenting the behavior should be sufficient.

Comment (mdhaber) [reply scipy/stats/_stats_py.py]: By *randomized* here, I'm referring to randomly resampled permutations as opposed to all possible permutations.

**Comment by mdhaber** [comment]
The test timeouts are happening in master. Thanks for taking a look @rlucas7! Perhaps it's easiest to compare against gh-14206, which I believe we thought was ready.

**Comment by tylerjereddy** [comment]
I can't quite tell if this is ready, so probably won't make it for 1.8.0rc2. It looks like an efficiency improvement that is not quite a bug, though debatable, so perhaps not super bad if it ends up in 1.8.1 or rc3 or whatever.

**Comment by mdhaber** [comment]
Thanks for taking a look @tylerjereddy. I don't personally consider it a bug, and it hasn't really been reported as such. But at the time we were working on `bootstrap`, there were strong opinions about the importance of batching, so I put on the backport-candidate label. It's fine to wait for merge by another stats maintainer. Thanks!

**Comment by mdhaber** [comment]
I think this is all ready to go again, @rlucas7. Not much has changed compared to gh-14206; just a few comments and a reference about the p-value. Thanks!

**Review by tupui — APPROVED** [review]
Thanks @mdhaber. Changes LGTM and it has already been reviewed. So no reason to hold this longer.

CI is just timing out as usual, so merging.

**Pull request merged**
